app/src/main/java/tech/relaycorp/gateway/data/disk/CargoStorage.kt

```diff
@@ -0,0 +1,51 @@
+package tech.relaycorp.gateway.data.disk
+
+import tech.relaycorp.gateway.common.Logging.logger
+import tech.relaycorp.gateway.common.Operation
+import tech.relaycorp.relaynet.cogrpc.readBytesAndClose
+import tech.relaycorp.relaynet.messages.Cargo
+import tech.relaycorp.relaynet.ramf.RAMFException
+import java.io.InputStream
+import javax.inject.Inject
+
+class CargoStorage
+@Inject constructor(
+    private val diskOperations: DiskOperations
+) {
+
+    suspend fun store(cargoStream: InputStream): Operation<String> {
+        val cargoBytes = cargoStream.readBytesAndClose()
+        val cargo = try {
+            Cargo.deserialize(cargoBytes)
+        } catch (exc: RAMFException) {
+            logger.warning("Malformed cargo received: ${exc.message}")
+            return Operation.Error(MalformedCargoException())
+        }
+
+        try {
+            cargo.validate()
```

Comment:
FYI: I added a TODO to https://github.com/relaycorp/relaynet-gateway-android/issues/19 to make sure to pass the gateway's certificate to this method, once we have the code to generate the gateway's certificate.